bpf/bpf_sock.c

```diff
@@ -367,33 +368,20 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 	if (sock4_skip_xlate(svc, orig_key.address))
 		return -EPERM;
 
-	if (lb4_svc_is_affinity(svc)) {
-		/* Note, for newly created affinity entries there is a
-		 * small race window. Two processes on two different
-		 * CPUs but the same netns may select different backends
-		 * for the same service:port. lb4_update_affinity_by_netns()
-		 * below would then override the first created one if it
-		 * didn't make it into the lookup yet for the other CPU.
-		 */
-		backend_id = lb4_affinity_backend_id_by_netns(svc, &id);
-		backend_from_affinity = true;
-
-		if (backend_id != 0) {
-			backend = __lb4_lookup_backend(backend_id);
-			if (!backend)
-				/* Backend from the session affinity no longer
-				 * exists, thus select a new one. Also, remove
-				 * the affinity, so that if the svc doesn't have
-				 * any backend, a subsequent request to the svc
-				 * doesn't hit the reselection again.
-				 */
-				backend_id = 0;
-		}
+	svc_has_affinity = lb4_svc_is_affinity(svc);
+	/* When service affinity is disabled, select a backend for
+	 * an unconnected UDP client based on the socket cookie.
+	 * This is to ensure that once a backend is selected for a client
+	 * socket, we continue to select the same backend even when
+	 * the service backends are updated.
+	 */
+	if (udp_only && !svc_has_affinity) {
```

Comment:
@borkmann I was in two minds whether to wrap this under `BPF_HAVE_SOCK_COOKIE` check. In the end, I trusted the compiler to do optimize the code path on older kernels where the BPF socket cookie helper isn't available.